server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -3666,7 +3677,12 @@ private class SnapshotTaskExecutor implements ClusterStateTaskExecutor<SnapshotT
         @Override
         public ClusterState execute(BatchExecutionContext<SnapshotTask> batchExecutionContext) throws Exception {
             final ClusterState state = batchExecutionContext.initialState();
-            final SnapshotShardsUpdateContext shardsUpdateContext = new SnapshotShardsUpdateContext(batchExecutionContext);
+            final SnapshotShardsUpdateContext shardsUpdateContext = new SnapshotShardsUpdateContext(
+                batchExecutionContext,
+                new RunOnce(
+                    () -> clusterService.getRerouteService().reroute("after shards snapshot update", Priority.NORMAL, ActionListener.noop())
+                )
```

Comment:
Why this have to be RunOnce?

Reply:
There's no value in queueing up a no-op listener per shard update on the reroute service, there could be thousands of them.

Reply:
Is it going to be called once first shard is completed or when all shards is restored from the snapshot?

Reply:
Neither, this is happening when shard snapshots complete.